storage/innobase/row/row0log.cc

```diff
@@ -205,6 +215,158 @@ struct row_log_t {
 				log operation */
 };
 
+struct crypt_info_t {
+
+	/** Encryption algorithm */
+	Encryption::Type	encryption_type;
+
+	/** Encryption key */
+	byte			encryption_key[ENCRYPTION_KEY_LEN];
+
+	/** Encryption key length*/
+	ulint			encryption_klen;
+};
+
+my_bool	srv_encrypt_online_alter_logs = FALSE;
+
+crypt_info_t crypt_info = crypt_info_t();
+
+/** Find out if temporary log files encrypted.
+@return true if temporary log file should be encrypted, false if not */
+bool
+log_tmp_is_encrypted()
+{
+	return (crypt_info.encryption_type != Encryption::NONE);
+}
+
+/** Check the row log encryption is enabled or not.
+It will enable the row log encryption. */
+void
+log_tmp_enable_encryption_if_set()
+{
+	if (srv_encrypt_online_alter_logs) {
+		if (crypt_info.encryption_type == Encryption::NONE) {
+			crypt_info.encryption_type = Encryption::AES;
+			crypt_info.encryption_klen = ENCRYPTION_KEY_LEN;
+			my_rand_buffer(crypt_info.encryption_key,
+				       ENCRYPTION_KEY_LEN);
+		}
+	}
+}
+
+/** Encrypt a temporary file block.
+@param[in]	src		block to encrypt or decrypt
+@param[in]	size		size of the block
+@param[out]	dst		destination block
+@param[in]	offs		offset to block
+@param[in]	space_id	tablespace id
+@return whether the operation succeeded */
+bool
+log_tmp_block_encrypt(
+/*==================*/
+	const byte*		src_block,
+	ulint			size,
+	byte*			dst_block,
+	os_offset_t		offs,
+	ulint			space_id)
+{
+	size_t dst_len;
+	byte iv[MY_AES_BLOCK_SIZE];
+
+	memset(iv, 0, MY_AES_BLOCK_SIZE);
+
+	mach_write_to_8(iv, space_id);
+	mach_write_to_8(iv + 8, offs);
+
+	int res = my_aes_crypt(MY_AES_ECB,
```

Comment:
Does `my_aes_crypt()` allow to do inplace encryption?

Reply:
Basically it is just a wrapper for `EVP_EncryptUpdate`, which does not allow `in` and `out` to partially overlap, but it does support in-place encryption. I am cheating here of course because I know that `iv` size equals to cipher block size and must be enough to accommodate the result, but I am doing it consciously.